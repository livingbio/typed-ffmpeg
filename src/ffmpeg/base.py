from typing import Any

from ffmpeg.schema import StreamType

from .dag.nodes import FilterableStream, FilterNode, InputNode, MergeOutputsNode, OutputNode, OutputStream
from .streams.audio import AudioStream
from .streams.av import AVStream
from .streams.video import VideoStream


def input(filename: str, **kwargs: Any) -> AVStream:
    """
    Input file URL (ffmpeg ``-i`` option)

    Args:
        filename: Input file URL
        **kwargs: ffmpeg's input file options

    Returns:
        Input stream

    Examples:
    ```py
    >>> input('input.mp4')
    <AVStream:input.mp4:0>
    ```
    """
    return InputNode(filename=filename, kwargs=tuple(kwargs.items())).stream()


def output(*streams: FilterableStream, filename: str, **kwargs: Any) -> OutputStream:
    """
    Output the streams to a file URL

    Args:
        *streams: the streams to output
        filename: the filename to output to
        **kwargs: the arguments for the output

    Returns:
        the output stream
    """
    return OutputNode(filename=filename, inputs=streams, kwargs=tuple(kwargs.items())).stream()


def merge_outputs(*streams: OutputStream) -> OutputStream:
    """
    Merge multiple output streams into one.

    Args:
        *streams: The output streams to merge.

    Returns:
        The merged output stream.
    """
    return MergeOutputsNode(inputs=streams).stream()


def vfilter(
    *streams: FilterableStream, name: str, input_typings: tuple[StreamType, ...] = (StreamType.video,), **kwargs: Any
) -> VideoStream:
    """
    Apply a custom video filter which has only one output to this stream

    Args:
        *streams: the streams to apply the filter to
        name: the name of the filter
        **kwargs: the arguments for the filter

    Returns:
        the output stream

    Note:
        This function is for custom filter which is not implemented in typed-ffmpeg
    """
    return FilterNode(
        name=name,
        inputs=streams,
        output_typings=(StreamType.video,),
        input_typings=input_typings,
        kwargs=tuple(kwargs.items()),
    ).video(0)


def afilter(
    *streams: FilterableStream, name: str, input_typings: tuple[StreamType, ...] = (StreamType.audio,), **kwargs: Any
) -> AudioStream:
    """
    Apply a custom audio filter which has only one output to this stream

    Args:
        *streams: the streams to apply the filter to
        name: the name of the filter
        **kwargs: the arguments for the filter

    Returns:
        the output stream

    Note:
        This function is for custom filter which is not implemented in typed-ffmpeg
    """
    return FilterNode(
        name=name,
        inputs=streams,
        output_typings=(StreamType.audio,),
        input_typings=input_typings,
        kwargs=tuple(kwargs.items()),
    ).audio(0)


def filter_multi_output(
    *streams: FilterableStream,
    name: str,
    input_typings: tuple[StreamType, ...] = (),
    output_tyings: tuple[StreamType, ...] = (),
    **kwargs: Any
) -> FilterNode:
    """
    Apply a custom filter which has multiple outputs to this stream

    Args:
        *streams: the streams to apply the filter to
        name: the name of the filter
        **kwargs: the arguments for the filter

    Returns:
        the FilterNode

    Note:
        This function is for custom filter which is not implemented in typed-ffmpeg
    """
    return FilterNode(
        name=name,
        kwargs=tuple(kwargs.items()),
        inputs=streams,
        input_typings=input_typings,
        output_typings=output_tyings,
    )
