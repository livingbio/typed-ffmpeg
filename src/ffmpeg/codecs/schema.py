from dataclasses import dataclass

from ..common.serialize import Serializable
from ..utils.frozendict import FrozenDict


@dataclass(frozen=True, kw_only=True)
class FFMpegCodecOption(Serializable):
    kwargs: FrozenDict[str, str | int | float | bool] = FrozenDict({})


@dataclass(frozen=True, kw_only=True)
class FFMpegEncoderOption(FFMpegCodecOption): ...


@dataclass(frozen=True, kw_only=True)
class FFMpegDecoderOption(FFMpegCodecOption): ...
