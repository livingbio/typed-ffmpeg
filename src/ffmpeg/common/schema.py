from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Literal


class StreamType(str, Enum):
    """
    The type of a stream. (audio or video)
    """

    audio = "audio"
    """it is an audio stream"""
    video = "video"
    """it is a video stream"""


class FFMpegFilterOptionType(str, Enum):
    boolean = "boolean"
    duration = "duration"
    color = "color"
    flags = "flags"
    dictionary = "dictionary"
    pix_fmt = "pix_fmt"
    int = "int"
    int64 = "int64"
    double = "double"
    float = "float"
    string = "string"
    video_rate = "video_rate"
    image_size = "image_size"
    rational = "rational"
    sample_fmt = "sample_fmt"
    binary = "binary"


class FFMpegFilterType(str, Enum):
    af = "af"
    asrc = "asrc"
    asink = "asink"
    vf = "vf"
    vsrc = "vsrc"
    vsink = "vsink"
    avsrc = "avsrc"
    avf = "avf"
    vaf = "vaf"


@dataclass(frozen=True, kw_only=True)
class FFMpegFilterOptionChoice:
    name: str
    help: str
    value: str | int
    flags: str | None = None


@dataclass(frozen=True, kw_only=True)
class FFMpegFilterOption:
    name: str
    alias: tuple[str, ...] = ()
    description: str
    type: FFMpegFilterOptionType
    min: str | None = None
    max: str | None = None
    default: bool | int | float | str | None = None
    required: bool = False
    choices: tuple[FFMpegFilterOptionChoice, ...] = ()
    flags: str | None = None


@dataclass(frozen=True, kw_only=True)
class FFMpegIOType:
    name: str | None = None
    type: StreamType


@dataclass(frozen=True, kw_only=True)
class FFMpegFilterDef:
    name: str

    typings_input: str | tuple[Literal["video", "audio"], ...] = ()
    typings_output: str | tuple[Literal["video", "audio"], ...] = ()


@dataclass(frozen=True, kw_only=True)
class FFMpegFilter:
    id: str | None = None

    name: str
    description: str
    ref: str | None = None

    # Flags
    is_support_slice_threading: bool | None = None
    is_support_timeline: bool | None = None
    is_support_framesync: bool | None = None
    is_support_command: bool | None = None
    is_filter_sink: bool | None = None
    is_filter_source: bool | None = None

    # IO Typing
    is_dynamic_input: bool = False
    is_dynamic_output: bool = False
    stream_typings_input: tuple[FFMpegIOType, ...] = ()
    stream_typings_output: tuple[FFMpegIOType, ...] = ()
    formula_typings_input: str | None = None
    formula_typings_output: str | None = None

    pre: tuple[tuple[str, str], ...] = ()
    options: tuple[FFMpegFilterOption, ...] = ()

    @property
    def pre_dict(self) -> dict[str, str]:
        return dict(self.pre)

    @property
    def to_def(self) -> FFMpegFilterDef:
        return FFMpegFilterDef(
            name=self.name,
            typings_input=self.formula_typings_input or tuple(k.type.value for k in self.stream_typings_input),
            typings_output=self.formula_typings_output or tuple(k.type.value for k in self.stream_typings_output),
        )

    @property
    def input_typings(self) -> set[StreamType]:
        if self.is_filter_source:
            return set()
        if not self.is_dynamic_input:
            return {i.type for i in self.stream_typings_input}
        else:
            assert self.formula_typings_input, f"{self.name} has no input"
            if "video" not in self.formula_typings_input:
                assert "audio" in self.formula_typings_input, f"{self.name} has no video input"
                return {StreamType.audio}
            elif "audio" not in self.formula_typings_input:
                assert "video" in self.formula_typings_input, f"{self.name} has no audio input"
                return {StreamType.video}
            assert (
                "video" in self.formula_typings_input and "audio" in self.formula_typings_input
            ), f"{self.name} has no video or audio input"
            return {StreamType.video, StreamType.audio}

    @property
    def output_typings(self) -> set[StreamType]:
        if self.is_filter_sink:
            return set()
        if not self.is_dynamic_output:
            return {i.type for i in self.stream_typings_output}
        else:
            assert self.formula_typings_output, f"{self.name} has no output"
            if "video" not in self.formula_typings_output:
                assert "audio" in self.formula_typings_output, f"{self.name} has no video output"
                return {StreamType.audio}
            elif "audio" not in self.formula_typings_output:
                assert "video" in self.formula_typings_output, f"{self.name} has no audio output"
                return {StreamType.video}
            assert (
                "video" in self.formula_typings_output and "audio" in self.formula_typings_output
            ), f"{self.name} has no video or audio output"
            return {StreamType.video, StreamType.audio}

    @property
    def filter_type(self) -> FFMpegFilterType:
        if self.is_filter_sink:
            assert len(self.input_typings) == 1
            if {StreamType.video} == self.input_typings:
                return FFMpegFilterType.vsink
            if {StreamType.audio} == self.input_typings:
                return FFMpegFilterType.asink
        elif self.is_filter_source:
            if {StreamType.video, StreamType.audio} == self.output_typings:
                return FFMpegFilterType.avsrc
            if {StreamType.video} == self.output_typings:
                return FFMpegFilterType.vsrc
            if {StreamType.audio} == self.output_typings:
                return FFMpegFilterType.asrc

        assert self.input_typings

        if self.input_typings == {StreamType.video}:
            if StreamType.audio in self.output_typings:
                return FFMpegFilterType.vaf
            if self.output_typings == {StreamType.video}:
                return FFMpegFilterType.vf

        if self.input_typings == {StreamType.audio}:
            if self.output_typings == {StreamType.audio}:
                return FFMpegFilterType.af
            if StreamType.video in self.output_typings:
                return FFMpegFilterType.avf

        if self.input_typings == {StreamType.video, StreamType.audio}:
            return FFMpegFilterType.avf

        raise ValueError(f"Unknown filter type for {self.name}")


class FFMpegOptionFlag(int, Enum):
    OPT_FUNC_ARG = 1 << 0
    """
    The OPT_TYPE_FUNC option takes an argument.
    Must not be used with other option types, as for those it holds:
    - OPT_TYPE_BOOL do not take an argument
    - all other types do
    """

    OPT_EXIT = 1 << 1
    """
    Program will immediately exit after processing this option
    """

    OPT_EXPERT = 1 << 2
    """
    Option is intended for advanced users. Only affects help output.
    """

    OPT_VIDEO = 1 << 3
    OPT_AUDIO = 1 << 4
    OPT_SUBTITLE = 1 << 5
    OPT_DATA = 1 << 6

    OPT_PERFILE = 1 << 7
    """
    The option is per-file (currently ffmpeg-only). At least one of OPT_INPUT or OPT_OUTPUT must be set when this flag is in use.
    """

    OPT_FLAG_OFFSET = 1 << 8
    """
    Option is specified as an offset in a passed optctx.
    Always use as OPT_OFFSET in option definitions.
    """

    OPT_OFFSET = OPT_FLAG_OFFSET | OPT_PERFILE
    """
    Option is to be stored in a SpecifierOptList.
    Always use as OPT_SPEC in option definitions.
    """
    OPT_FLAG_SPEC = 1 << 9
    """
    Option is to be stored in a SpecifierOptList.
    Always use as OPT_SPEC in option definitions.
    """

    OPT_SPEC = OPT_FLAG_SPEC | OPT_OFFSET
    """
    Option applies per-stream (implies OPT_SPEC).
    """
    OPT_FLAG_PERSTREAM = 1 << 10
    """
    Option applies per-stream (implies OPT_SPEC).
    """
    OPT_PERSTREAM = OPT_FLAG_PERSTREAM | OPT_SPEC

    OPT_INPUT = 1 << 11
    """
    ffmpeg-only - specifies whether an OPT_PERFILE option applies to input, output, or both.
    """
    OPT_OUTPUT = 1 << 12
    """
    ffmpeg-only - specifies whether an OPT_PERFILE option applies to input, output, or both.
    """

    OPT_HAS_ALT = 1 << 13
    """
    This option is a "canonical" form, to which one or more alternatives exist. These alternatives are listed in u1.names_alt.
    """
    OPT_HAS_CANON = 1 << 14
    """
    This option is an alternative form of some other option, whose name is stored in u1.name_canon
    """


class FFMpegOptionType(str, Enum):
    OPT_TYPE_FUNC = "OPT_TYPE_FUNC"
    OPT_TYPE_BOOL = "OPT_TYPE_BOOL"
    OPT_TYPE_STRING = "OPT_TYPE_STRING"
    OPT_TYPE_INT = "OPT_TYPE_INT"
    OPT_TYPE_INT64 = "OPT_TYPE_INT64"
    OPT_TYPE_FLOAT = "OPT_TYPE_FLOAT"
    OPT_TYPE_DOUBLE = "OPT_TYPE_DOUBLE"
    OPT_TYPE_TIME = "OPT_TYPE_TIME"


@dataclass(frozen=True, kw_only=True)
class FFMpegOption:
    name: str
    type: FFMpegOptionType
    flags: int
    help: str
    argname: str | None = None
    canon: str | None = None

    @property
    def is_input_option(self) -> bool:
        return bool(self.flags & FFMpegOptionFlag.OPT_INPUT)

    @property
    def is_output_option(self) -> bool:
        return bool(self.flags & FFMpegOptionFlag.OPT_OUTPUT)

    @property
    def is_global_option(self) -> bool:
        return not self.is_input_option and not self.is_output_option and not (self.flags & FFMpegOptionFlag.OPT_EXIT)

    @property
    def is_support_stream_specifier(self) -> bool:
        return bool(self.flags & FFMpegOptionFlag.OPT_SPEC)
