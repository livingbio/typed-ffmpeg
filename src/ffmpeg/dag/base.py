from __future__ import annotations

from abc import ABC, abstractmethod, abstractproperty
from dataclasses import dataclass
from functools import cached_property
from typing import Iterable, Sequence

from ..schema import StreamType
from ..utils.dag import is_dag
from ..utils.typing import override


class _DAGContext(ABC):
    """
    An abstract class context for DAG.
    """

    @abstractmethod
    def get_node_label(self, node: Node) -> str:
        """
        Get the label of the node.

        Args:
            node: The node to get the label of.

        Returns:
            The label of the node.
        """
        raise NotImplementedError()

    @abstractmethod
    def get_outgoing_streams(self, node: Node) -> Iterable[Stream]:
        """
        Extract all node's outgoing streams from the given set of streams, Because a node only know its incoming streams.

        Args:
            node: The node to get the outgoing streams of.

        Returns:
            The outgoing streams of the node.
        """
        raise NotImplementedError()


class DummyDAGContext(_DAGContext):
    """A dummy DAG context that does not do anything"""

    @override
    def get_node_label(self, node: Node) -> str:
        return str(node)

    @override
    def get_outgoing_streams(self, node: Node) -> Iterable[Stream]:
        return []


empty_dag_context = DummyDAGContext()


@dataclass(frozen=True, kw_only=True)
class HashableBaseModel:
    @cached_property
    def hex(self) -> str:
        return hex(abs(hash(self)))[2:]

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({self.hex})"


@dataclass(frozen=True, kw_only=True)
class Stream(HashableBaseModel):
    """
    A 'Stream' represents a sequence of data flow in the Directed Acyclic Graph (DAG).

    Note:
        Each stream in the DAG is a sequence of operations that transforms the data from its input form to its output form. The stream is an essential component of the DAG, as it defines the order and the nature of the operations that are performed on the data.
    """

    node: Node
    """
    Represents the node that the stream is connected to in the upstream direction.

    Note:
        In the context of a data stream, the 'upstream' refers to the source of the data, or where the data is coming from. Therefore, the 'upstream node' is the node that is providing the data to the current stream.
    """
    selector: StreamType | None = None
    """
    Represents the type of the stream.

    Notes:
        See Also: [Stream specifiers](https://ffmpeg.org/ffmpeg.html#Stream-specifiers-1) `stream_type`
    """

    index: int | None = None  # the nth child of the node
    """
    Represents the index of the stream in the node's output streams.

    Note:
        See Also: [Stream specifiers](https://ffmpeg.org/ffmpeg.html#Stream-specifiers-1) `stream_index`
    """


@dataclass(frozen=True, kw_only=True)
class Node(HashableBaseModel, ABC):
    """
    A 'Node' represents a single operation in the Directed Acyclic Graph (DAG).

    Note:
        Each node in the DAG represents a single operation that transforms the data from its input form to its output form. The node is an essential component of the DAG, as it defines the nature of the operations that are performed on the data.
    """

    args: tuple[str, ...] = ()
    kwargs: tuple[tuple[str, str | int | float | bool], ...] = ()

    @abstractproperty
    def incoming_streams(self) -> Sequence["Stream"]:
        """
        Return the incoming streams of this node.

        Returns:
            The incoming streams of this node.
        """
        raise NotImplementedError()

    def __post_init__(self) -> None:
        # Validate the DAG
        passed = set()
        nodes = [self]
        output = {}

        while nodes:
            node = nodes.pop()

            if node in passed:
                continue
            passed.add(node)

            nodes.extend(k.node for k in node.incoming_streams)

            output[str(node)] = set(str(k.node) for k in node.incoming_streams)

        if not is_dag(output):
            raise ValueError(f"Graph is not a DAG: {output}")

    @abstractmethod
    def get_args(self, context: _DAGContext = empty_dag_context) -> list[str]:
        """
        Get the arguments of the node.

        Args:
            context: The DAG context. Defaults to empty_dag_context.

        Returns:
            The arguments of the node.
        """
        raise NotImplementedError()
