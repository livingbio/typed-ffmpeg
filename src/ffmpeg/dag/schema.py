from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, replace
from functools import cached_property
from typing import TYPE_CHECKING

from .utils import is_dag

if TYPE_CHECKING:
    from .context import DAGContext


@dataclass(frozen=True, kw_only=True)
class HashableBaseModel:
    @cached_property
    def hex(self) -> str:
        return hex(abs(hash(self)))[2:]


@dataclass(frozen=True, kw_only=True)
class Stream(HashableBaseModel):
    """
    A 'Stream' represents a sequence of data flow in the Directed Acyclic Graph (DAG).

    Note:
        Each stream in the DAG is a sequence of operations that transforms the data from its input form to its output form. The stream is an essential component of the DAG, as it defines the order and the nature of the operations that are performed on the data.
    """

    node: Node
    """
    Represents the node that the stream is connected to in the upstream direction.

    Note:
        In the context of a data stream, the 'upstream' refers to the source of the data, or where the data is coming from. Therefore, the 'upstream node' is the node that is providing the data to the current stream.
    """

    index: int | None = None  # the nth child of the node
    """
    Represents the index of the stream in the node's output streams.

    Note:
        See Also: [Stream specifiers](https://ffmpeg.org/ffmpeg.html#Stream-specifiers-1) `stream_index`
    """

    def view(self, format: str = "png") -> str:
        from ..utils.view import view

        return view(self.node, format=format)


@dataclass(frozen=True, kw_only=True)
class Node(HashableBaseModel, ABC):
    """
    A 'Node' represents a single operation in the Directed Acyclic Graph (DAG).

    Note:
        Each node in the DAG represents a single operation that transforms the data from its input form to its output form. The node is an essential component of the DAG, as it defines the nature of the operations that are performed on the data.
    """

    args: tuple[str, ...] = ()
    kwargs: tuple[tuple[str, str | int | float | bool], ...] = ()
    inputs: tuple[Stream, ...] = ()

    def __post_init__(self) -> None:
        # Validate the DAG
        passed = set()
        nodes = [self]
        output = {}

        while nodes:
            node = nodes.pop()

            if node in passed:
                continue
            passed.add(node)

            nodes.extend(k.node for k in node.inputs)

            output[node.hex] = set(k.node.hex for k in node.inputs)

        if not is_dag(output):
            raise ValueError(f"Graph is not a DAG: {output}")  # pragma: no cover

    @abstractmethod
    def get_args(self, context: DAGContext = None) -> list[str]:
        """
        Get the arguments of the node.

        Args:
            context: The DAG context. Defaults to empty_dag_context.

        Returns:
            The arguments of the node.
        """

    def repr(self) -> str:
        """
        Get the representation of the node.

        Returns:
            The representation of the node.
        """
        return repr(self)

    def replace(self, old_node: Node, new_node: Node) -> Node:
        """
        Replace the old node with the new node.

        Args:
            old_node: The old node to replace.
            new_node: The new node to replace with.

        Returns:
            The new node.
        """
        if self == old_node:
            return new_node

        inputs = []

        for stream in self.inputs:
            new_stream_node = stream.node.replace(old_node, new_node)

            if new_stream_node != stream.node:
                # need to create a new stream
                new_stream = replace(stream, node=new_stream_node)
                inputs.append(new_stream)
            else:
                inputs.append(stream)

        return replace(self, inputs=tuple(inputs))

    @property
    def max_depth(self) -> int:
        """
        Get the maximum depth of the node.

        Returns:
            The maximum depth of the node.
        """
        return max((i.node.max_depth for i in self.inputs), default=0) + 1

    @property
    def upstream_nodes(self) -> set[Node]:
        """
        Get all upstream nodes of the node.
        Returns:
            The upstream nodes of the node.
        """
        output = {self}
        for input in self.inputs:
            output |= input.node.upstream_nodes

        return output
