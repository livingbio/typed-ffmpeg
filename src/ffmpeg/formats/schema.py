from dataclasses import dataclass

from ..common.serialize import Serializable
from ..utils.frozendict import FrozenDict


@dataclass(frozen=True, kw_only=True)
class FFMpegFormatOption(Serializable):
    kwargs: FrozenDict[str, str | int | float | bool] = FrozenDict({})


@dataclass(frozen=True, kw_only=True)
class FFMpegMuxerOption(FFMpegFormatOption): ...


@dataclass(frozen=True, kw_only=True)
class FFMpegDemuxerOption(FFMpegFormatOption): ...
