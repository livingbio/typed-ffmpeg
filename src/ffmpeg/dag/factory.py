import re
from typing import Any

from ..common.schema import FFMpegFilterDef, StreamType
from ..schema import Auto
from ..utils.run import _to_tuple
from .nodes import FilterableStream, FilterNode


def filter_node_factory(filter: FFMpegFilterDef, *inputs: FilterableStream, **kwargs: Any) -> FilterNode:
    for k, v in kwargs.items():
        if isinstance(v, Auto):
            kwargs[k] = eval(v, {"StreamType": StreamType, "re": re, **kwargs, "streams": inputs})

    if isinstance(filter.typings_input, str):
        input_typings = tuple(eval(filter.typings_input, {"StreamType": StreamType, "re": re, **kwargs}))
    else:
        input_typings = tuple(StreamType.video if k == "video" else StreamType.audio for k in filter.typings_input)

    if isinstance(filter.typings_output, str):
        output_typings = tuple(eval(filter.typings_output, {"StreamType": StreamType, "re": re, **kwargs}))
    else:
        output_typings = tuple(StreamType.video if k == "video" else StreamType.audio for k in filter.typings_output)

    return FilterNode(
        name=filter.name,
        input_typings=input_typings,
        output_typings=output_typings,
        inputs=inputs,
        kwargs=_to_tuple(kwargs),
    )
