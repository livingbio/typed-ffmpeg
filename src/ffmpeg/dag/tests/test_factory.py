from syrupy.assertion import SnapshotAssertion

from ...base import input
from ...common.schema import FFMpegFilterDef
from ...schema import Auto
from ..factory import filter_node_factory


def test_filter_node_factory(snapshot: SnapshotAssertion) -> None:
    in_file = input("foo.mp4")

    f = filter_node_factory(
        FFMpegFilterDef(name="foo", typings_input="[StreamType.video]*n"), **{"n": Auto("len(streams)")}
    )
    assert snapshot == f
    assert f.kwargs == tuple({"n": 0}.items())

    f = filter_node_factory(
        FFMpegFilterDef(name="foo", typings_input="[StreamType.video]*n"), in_file.video, **{"n": Auto("len(streams)")}
    )
    assert snapshot == f
    assert f.kwargs == tuple({"n": 1}.items())

    f = filter_node_factory(
        FFMpegFilterDef(name="foo", typings_input="[StreamType.video]*n"),
        in_file.video,
        in_file.video,
        **{"n": Auto("len(streams)")}
    )
    assert snapshot == f
    assert f.kwargs == tuple({"n": 2}.items())
