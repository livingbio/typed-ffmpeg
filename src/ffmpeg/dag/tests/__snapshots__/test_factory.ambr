# serializer version: 1
# name: test_filter_node_factory
  FilterNode(kwargs=(('n', 0),), inputs=(), name='foo', input_typings=(), output_typings=())
# ---
# name: test_filter_node_factory.1
  FilterNode(kwargs=(('n', 1),), inputs=(VideoStream(node=InputNode(kwargs=(), inputs=(), filename='foo.mp4'), index=None),), name='foo', input_typings=(<StreamType.video: 'video'>,), output_typings=())
# ---
# name: test_filter_node_factory.2
  FilterNode(kwargs=(('n', 2),), inputs=(VideoStream(node=InputNode(kwargs=(), inputs=(), filename='foo.mp4'), index=None), VideoStream(node=InputNode(kwargs=(), inputs=(), filename='foo.mp4'), index=None)), name='foo', input_typings=(<StreamType.video: 'video'>, <StreamType.video: 'video'>), output_typings=())
# ---
